### DISCUSSION TITLE
Nuxt 3 > Express Server CORS error (on same machine)

### DISCUSSION THREAD

**Discussion by braitsch** [nuxt/nuxt#16985] [General]
Hi, I have simple Nuxt 3 app that calls an API endpoint on an Express server (running on the same machine) and I'm getting a CORS error (in production only) that I can't figure out how to resolve.

My Nuxt app is running on port 8080 via:
``PORT=8080 node .output/server/index.mjs``
My understanding is this is [an acceptable way to run a Nuxt 3 app in production.](https://v3.nuxtjs.org/guide/deploy/node-server)

The Nuxt app makes a GET call to the Express app which is running on port 8081:

	const response = await $fetch('http://localhost:8081/api/test')
	console.log('response:', response);

This works fine when running locally but as soon as I put both apps on a VPS I get a CORS error:
````
Access to fetch at 'http://localhost:8081/api/test' from origin 'http://147.182.204.161:8080' has been blocked by CORS policy: The request client is not a secure context and the resource is in more-private address space `local`.
````
I get the same error if I used a fully qualified domain name instead of the IP address.

I tried adding fully open CORS to the Express app and it didn't make a difference. In fact I get the CORS error regardless of whether the Express app is running! So it seems I'm missing something in the Nuxt app to allow it to communicate w/ the Express server on the same machine.

Any help is very much appreciated!

**Comment by john-prutton** [comment]
In my limited understanding, it seems for your current setup to work you would need to be running your Nuxt app on your VPS at 8080 and your Express app on your local machine at 8081. This is obviously not what you want. Have you tried replacing localhost with the IP of your VPS inside the $fetch request?

`const response = await $fetch('http://<VPS_IP>:8081/api/test)`

I think a better way, if someone could please correct me otherwise, would be to use the `~/server/api/` routes as a proxy to pass requests to your Express app. Maybe then you could use `localhost:8081`, since, again to my limited knowledge, the Nuxt backend running on your VPS can make the request to its localhost. Really not too sure.

Another option, which I know the least about, is to use Express inside `~/server/`, maybe inside a nitro plugin (see nuxt/framework#5525 for an example of a nitro plugin). Then, perhaps, you can use `~/server/api/` routes to call/define Express routes?

Again, I hope someone can correct any mistakes I've made.  

(Also, it doesn't seem like a good idea to expose your VPS IP address in a public post; even I tried to SSH.)
